# Check duplicates

In [6]:
import pandas as pd
import sqlite3
import datetime
from Onehousing.cleaning import OneHousingDataCleaner 

pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', None)

In [20]:
with sqlite3.connect('output/real_estate.db') as conn:
    bds_raw = pd.read_sql('SELECT * FROM bds_raw', conn)
    onehousing_raw = pd.read_sql('SELECT * FROM onehousing_raw', conn)
    cleaned = pd.read_sql('SELECT * FROM cleaned', conn)

In [ ]:
print(f'Number of duplicated rows in bds_raw: {bds_raw.duplicated().sum()} / {bds_raw.shape[0]}')
print(f'Number of duplicated rows in onehousing_raw: {onehousing_raw.duplicated().sum()} / {onehousing_raw.shape[0]}')
print(f'Number of duplicated rows in cleaned: {cleaned.duplicated().sum()} / {cleaned.shape[0]}')

Number of duplicated rows in bds_raw: 0 / 127725
Number of duplicated rows in onehousing_raw: 0 / 4478
Number of duplicated rows in cleaned: 0 / 15794


In [23]:
cleaned[cleaned["Web"] == "Onehousing"]

,Tỉnh/Thành phố,Thành phố/Quận/Huyện/Thị xã,Xã/Phường/Thị trấn,Đường phố,Chi tiết,Nguồn thông tin,Tình trạng giao dịch,Thời điểm giao dịch/rao bán,Thông tin liên hệ,Giá rao bán/giao dịch,Giá ước tính,Loại đơn giá (đ/m2 hoặc đ/m ngang),Đơn giá đất,Lợi thế kinh doanh,Số tầng công trình,Tổng diện tích sàn,Đơn giá xây dựng,Năm xây dựng,Chất lượng còn lại,Diện tích đất (m2),Kích thước mặt tiền (m),Kích thước chiều dài (m),Số mặt tiền tiếp giáp,Hình dạng,Độ rộng ngõ/ngách nhỏ nhất (m),Khoảng cách tới trục đường chính (m),Mục đích sử dụng đất,Yếu tố khác,Tọa độ (vĩ độ),Tọa độ (kinh độ),Hình ảnh của bài đăng,Web
11681,Thành phố Hà Nội,Quận Hà Đông,Phường Vạn Phúc,Phố Lụa,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Pho-Lua-100m--P--Van-Phuc--Q--Ha-Dong--TP--Ha-Noi.ULHEKO,Đang giao dịch,20/01/2026,nanval,9.200000e+09,9016000000,đ/m2,nanval,nanval,5.0,169.5,4000000.0,nanval,0.85,33.9,7.80,4.35,1,Chữ nhật,2.0,100.0,Đất ở,nanval,21.00101,105.866667,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/RESIDENTIAL/355c72a5-37b5-4b62-819a-f1f34e64d071/2c9427839a4a2959019b2c413fa75633_CLONE/LIVING_ROOM/scaled_1000040570_1765975397648.jpg,Onehousing
11682,Thành phố Hà Nội,Quận Bắc Từ Liêm,Phường Xuân Đỉnh,Xuân Đỉnh,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Xuan-Dinh-50m--P--Xuan-Dinh--Q--Bac-Tu-Liem--TP--Ha-Noi.SKUOOI,Đang giao dịch,20/01/2026,nanval,1.700000e+10,16660000000,đ/m2,nanval,nanval,6.0,241.8,4000000.0,nanval,0.85,40.3,4.10,9.83,1,Chữ nhật,4.0,50.0,Đất ở,nanval,21.00101,105.866667,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/RESIDENTIAL/eb81035f-a264-4aed-932c-e524bac4e5e5/2c9427839a4a2959019ba140fe0c7de0_CLONE/LIVING_ROOM/image_picker_0225B86C-1140-496C-9131-6DC827356E26-75883-00001587015572D0_1767937910623.jpg,Onehousing
11683,Thành phố Hà Nội,Quận Hai Bà Trưng,Phường Vĩnh Tuy,Kim Ngưu,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Kim-Nguu-100m--P--Vinh-Tuy--Q--Hai-Ba-Trung--TP--Ha-Noi.YWGUVT,Đang giao dịch,20/01/2026,nanval,5.200000e+09,5096000000,đ/m2,nanval,nanval,4.5,115.2,4000000.0,nanval,0.85,25.6,1.20,21.33,1,Chữ nhật,2.0,100.0,Đất ở,nanval,21.00101,105.866667,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/RESIDENTIAL/62ea34f7-62ce-4f11-92c8-cd8d8c9ca823/2c9427839a4a2959019b30c1bfac58c6_CLONE/LIVING_ROOM/scaled_1000060304_1766050118802.jpg,Onehousing
11684,Thành phố Hà Nội,Quận Hà Đông,Phường Hà Cầu,Hà Trì,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Ha-Tri-20m--P--Ha-Cau--Q--Ha-Dong--TP--Ha-Noi.IIWMDZ,Đang giao dịch,20/01/2026,nanval,1.850000e+10,18130000000,đ/m2,nanval,nanval,4.0,306.0,4000000.0,nanval,0.85,76.5,5.00,15.30,1,Chữ nhật,6.0,20.0,Đất ở,nanval,21.00101,105.866667,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/RESIDENTIAL/8ebf97a2-39ac-491a-a797-aad2e820c127/2c9427839a4a2959019b48e5eb8d5f0c_CLONE/LIVING_ROOM/scaled_1000060911_1766455083168.jpg,Onehousing
11685,Thành phố Hà Nội,Quận Đống Đa,Phường Thịnh Quang,Thái Thịnh,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Thai-Thinh-30m--P--Thinh-Quang--Q--Dong-Da--TP--Ha-Noi.NKHHRZ,Đang giao dịch,20/01/2026,nanval,1.950000e+10,19110000000,đ/m2,nanval,nanval,6.0,213.6,4000000.0,nanval,0.85,35.6,3.40,10.47,1,Chữ nhật,3.0,30.0,Đất ở,nanval,21.00101,105.866667,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/RESIDENTIAL/8995d755-c3d8-4b33-9b79-39c28c273d4f/2c9427839a4a2959019ba138a6847ddd_CLONE/LIVING_ROOM/image_picker_2E3A0736-3D43-4694-BD18-B10DAE346EF8-88338-00002DF9C82EA189_1767937353914.jpg,Onehousing
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15789,Thành phố Hà Nội,Quận Đống Đa,Phường Thịnh Quang,Tây Sơn,Mặt ngõ,https://onehousing.vn/bds/Nha-mat-ngo-cach-Tay-Son-50m--P--Thinh-Quang--Q--Dong-Da--TP--Ha-Noi.FAGEYE,Đang giao d

# Sửa lại cleaning rules cho Onehousing 

### 1. Scrape nhà liền kề 

In [14]:
# TEST SCRAPING VỚI 1 TRANG 

from Onehousing.fetch_listings import extract_listing_details
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager

def init_driver(headless=True):
    chrome_options = Options()
    
    if headless:
        chrome_options.add_argument("--headless=new")
    
    chrome_options.add_argument("--start-maximized")
    chrome_options.add_argument("--disable-blink-features=AutomationControlled")
    chrome_options.add_argument("--disable-gpu")
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
    
    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=chrome_options)
    
    driver.set_page_load_timeout(30)
    
    return driver

if __name__ == "__main__":
    
    TARGET_URL = "https://onehousing.vn/bds/Biet-thu-Hai-Au---Vinhomes-Ocean-Park-2---The-Empire.RMI2BR"  
    
    driver = init_driver(headless=True)  
    
    try:
        data = extract_listing_details(driver, TARGET_URL)
        
        if data:
            print("=== SCRAPED DATA ===")
            for key, value in data.items():
                print(f"{key}: {value}")
        else:
            print("Không lấy được dữ liệu.")
    
    finally:
        driver.quit()
        print("Driver đã đóng.")

# Hàm cho nhà đất chưa scrape được phần description cho nhà liền kề 

=== SCRAPED DATA ===
property_url: https://onehousing.vn/bds/Biet-thu-Hai-Au---Vinhomes-Ocean-Park-2---The-Empire.RMI2BR
listing_title: Biệt thự Hải Âu - Vinhomes Ocean Park 2 - The Empire
property_id: RMI2BR
total_price: 27.2 tỷ
unit_price: 168 triệu/m²
alley_width: Tình trạng pháp lý
Giấy chứng nhận
image_url: https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/1b41c1cb-a2d9-4c1b-8eee-53a2c210b736/344e3af2-a757-ca31-5236-c21194693828_CLONE/LIVING_ROOM/image_picker_B2B4A7EB-014F-4F05-BE26-50EE7E63E117-1669-000002400ECC174A_1750130319943.jpg
city: T. Hưng Yên
district: H. Văn Giang
features: Hướng mặt tiền: Đông Nam - 9m; Số tầng: 4; Diện tích: 162 m²; Số WC: 4; Nội thất: Xây thô; Tình trạng sử dụng: Để trống
latitude: 21.00100964213596
longitude: 105.8666666648782
property_description: Bán Biệt thự Hải Âu - Vinhomes Ocean Park 2 - The Empire
Biệt thự có vị trí đẹp, giá ưu đãi từ chủ nhà, pháp lý minh bạch, sẵn sàng giao dịch. Xem nhà ngay....


In [13]:
# Test scraping phần description cho nhà liền kề
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException
import time
from bs4 import BeautifulSoup

import re 

driver = init_driver(headless=True)

driver.get("https://onehousing.vn/bds/Biet-thu-Hai-Au---Vinhomes-Ocean-Park-2---The-Empire.RMI2BR")

def extract_clean_description(driver, wait_time=3):
    # Chờ trang render xong
    time.sleep(wait_time)

    soup = BeautifulSoup(driver.page_source, "html.parser")

    container = soup.select_one("div.max-h-\\[70px\\].overflow-hidden")

    if not container:
        return None

    html = str(container)

    # Convert block tags thành newline
    html = re.sub(r'<br\s*/?>', '\n', html)
    html = re.sub(r'</p>|</div>|</li>', '\n', html)

    # Remove tất cả HTML tags
    text = re.sub(r'<[^>]+>', '', html)

    # Clean spacing
    text = re.sub(r'\n+', '\n', text)
    text = re.sub(r'[ \t]+', ' ', text)

    return text.strip()

hidden_text = extract_clean_description(driver)
print(hidden_text)

Bán Biệt thự Hải Âu - Vinhomes Ocean Park 2 - The Empire
Biệt thự có vị trí đẹp, giá ưu đãi từ chủ nhà, pháp lý minh bạch, sẵn sàng giao dịch. Xem nhà ngay.
Thông tin mô tả Biệt thự:
Biệt thự có diện tích 162m² xây 4 tầng với mặt tiền hướng Đông Nam...
Tình trạng pháp lý: Giấy chứng nhận...
Biệt thự nằm trong phân khu Hải Âu, được thừa hưởng mọi tiện ích của dự án Vinhomes Ocean Park 2 - The Empire, tạo nên một cuộc sống vô cùng thoải mái và tiện nghi cho cư dân. Liên hệ với OneHousing để xem nhà ngay!
Các hạn chế về quyền sở hữu: Đang cập nhật


In [2]:
# Test scraping toàn bộ URLs của nhà liền kề

from Onehousing.fetch_urls import * 
import time
from typing import List, Set, Optional
import requests
from bs4 import BeautifulSoup


def build_search_url(original_url: Optional[str] = None) -> str:
    if original_url:
        # Nếu URL chưa có dấu ?, thêm vào
        if not original_url.endswith("?"):
            if "?" not in original_url:
                original_url = original_url.rstrip("/") + "?"
            else:
                original_url += "&"
        return original_url

    return get_search_url()


def scrape_all_listing_urls(
    start_page: int = 1,
    max_pages: int = 100,
    user_agent: str = None,
    delay: float = 1.0,
    original_url: Optional[str] = None,
) -> List[str]:

    base_search_url = build_search_url(original_url)
    all_urls: Set[str] = set()

    if not user_agent:
        user_agent = (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/123.0.0.0 Safari/537.36"
        )

    headers = {"User-Agent": user_agent}

    for page in range(start_page, start_page + max_pages):
        url = f"{base_search_url}page={page}"
        print(f"[Scraping] {url}")

        try:
            response = requests.get(url, headers=headers, timeout=10)

            if response.status_code != 200:
                print(f"[Fetch URL] HTTP {response.status_code} on page {page}")
                break

            urls = parse_listing_urls(response.text)

            if not urls:
                print(f"[Stop] No URLs found on page {page}")
                break

            before_count = len(all_urls)
            all_urls.update(urls)
            after_count = len(all_urls)

            print(f"[Info] Found {len(urls)} | Total unique: {after_count}")

            time.sleep(delay)

        except Exception as e:
            print(f"[Error] Page {page}: {e}")
            break

    return list(all_urls)


if __name__ == "__main__":
    custom_url = "https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?"
    urls_custom = scrape_all_listing_urls(
        max_pages=5,
        original_url=custom_url
    )

    print(f"Custom URLs: {len(urls_custom)}")

[Scraping] https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?page=1
[Info] Found 20 | Total unique: 20
[Scraping] https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?page=2
[Info] Found 20 | Total unique: 40
[Scraping] https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?page=3
[Info] Found 20 | Total unique: 60
[Scraping] https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?page=4
[Info] Found 2 | Total unique: 62
[Scraping] https://onehousing.vn/ban-lien-ke-shophouse-biet-thu?page=5
[Stop] No URLs found on page 5
Custom URLs: 62


In [ ]:
# Scrape toàn bộ details của nhà liền kề và construct lại thành một DataFrame 
# Stale element exception: 
from tqdm import tqdm
from Onehousing.fetch_listings import extract_listing_details

def scrape_all_listing_details(driver, list_urls, delay: float = 1.5):
    all_data = []

    for url in tqdm(list_urls, desc="Scraping listing details"):
        try:
            data = extract_listing_details(driver, url)

            if data:
                all_data.append(data)

            time.sleep(delay)

        except Exception as e:
            print(f"[Error] Failed at {url}: {e}")
            continue

    df = pd.DataFrame(all_data)

    return df

if __name__ == "__main__":
    driver = init_driver(headless=True)
    df = scrape_all_listing_details(driver, urls_custom)

    print("Total records:", len(df))

Scraping listing details: 100%|██████████| 62/62 [10:53<00:00, 10.54s/it]

Total records: 62
                                                                                      property_url  \
0  https://onehousing.vn/bds/Shophouse-Vinh-Thien-Duong---Vinhomes-Ocean-Park-3---The-Crown.49Q1RH   
1         https://onehousing.vn/bds/Nha-lien-ke-Cha-La---Vinhomes-Ocean-Park-2---The-Empire.NERYSZ   
2          https://onehousing.vn/bds/Shophouse-Pho-Bien---Vinhomes-Ocean-Park-3---The-Crown.MKMPWW   
3         https://onehousing.vn/bds/Nha-lien-ke-San-Ho---Vinhomes-Ocean-Park-2---The-Empire.RUVX4T   
4  https://onehousing.vn/bds/Shophouse-Vinh-Thien-Duong---Vinhomes-Ocean-Park-3---The-Crown.X8AZTB   

                                                    listing_title property_id  \
0  Shophouse Vịnh Thiên Đường - Vinhomes Ocean Park 3 - The Crown      49Q1RH   
1         Nhà liền kề Chà Là - Vinhomes Ocean Park 2 - The Empire      NERYSZ   
2          Shophouse Phố Biển - Vinhomes Ocean Park 3 - The Crown      MKMPWW   
3         Nhà liền kề San Hô - Vinhomes Ocean

In [24]:
df.head(1)

,property_url,listing_title,property_id,total_price,unit_price,alley_width,image_url,city,district,features,latitude,longitude,property_description
0,https://onehousing.vn/bds/Shophouse-Vinh-Thien-Duong---Vinhomes-Ocean-Park-3---The-Crown.49Q1RH,Shophouse Vịnh Thiên Đường - Vinhomes Ocean Park 3 - The Crown,49Q1RH,31.6 tỷ,247 triệu/m²,Tình trạng pháp lý\nHĐMB,https://rs.onehousing.vn/fast/320x0/filters:quality(95):format(webp)/cdn1.onehousing.vn/media/e5d7926f-5a11-46fe-a9be-c106c861fb95/eab04c34-0ef9-de24-3d70-061d66f0596f_CLONE/LIVING_ROOM/image_picker_41722ACA-A2E6-4CDD-A8E0-10E5FC74FBC9-6517-000005BE6D95FB18_1733457575982.jpg,T. Hưng Yên,H. Văn Giang,Hướng mặt tiền: Đông Nam - 8m; Số tầng: 5; Diện tích: 128 m²; Số WC: 4; Nội thất: Xây thô; Tình trạng sử dụng: Để trống,21.00100964213596,105.8666666648782,"Bán Shophouse Vịnh Thiên Đường - Vinhomes Ocean Park 3 - The Crown\nShophouse có vị trí đẹp, giá ưu đãi từ chủ nhà, pháp lý minh bạch, sẵn sàng giao dịch. Xem nhà ngay....\nThông tin mô tả Shophouse:\nShophouse có diện tích 128m² xây 5 tầng với mặt tiền hướng Đông Nam...\nTình trạng pháp lý: HĐMB...\nShophouse nằm trong phân khu Vịnh Thiên Đường, được thừa hưởng mọi tiện ích của dự án Vinhomes Ocean Park 3 - The Crown, tạo nên một cuộc sống vô cùng thoải mái và tiện nghi cho cư dân. Liên hệ với OneHousing để xem nhà ngay!\nCác hạn chế về quyền sở hữu: Đang cập nhật\n"


### 2. Đơn giá đất

Nếu mảnh này chỉ bán đất: Đơn giá đất = Giá ước tính / Diện tích đất

Nếu mảnh này bao gồm nhà (tầng): Lấy giá đất chia cho giá xây dựng ước tính

1. Giá xây dựng = Đơn giá xây dựng * Tổng diện tích sàn * Chất lượng còn lại

2. Đơn giá đất = Giá ước tính - Giá xây dựng

### 3. Tổng diện tích sàn 

1. Nếu trích xuất được các cụm diện tích sàn, diện tích sử dụng thì sẽ lưu nó thành tổng diện tích sàn
2. Nếu không có những cụm trên: Tổng diện tích sàn = diện tích xây dựng * số tầng công trình. Diện tích xây dựng được xác định như sau:
- Nếu tiêu đề bài đăng có 1 trong các cụm từ ""biệt thự, nhà liền kề, shophouse"":
    * Biệt thự: Diện tích xây dựng = 0.56 * diện tích đất
    * Liền kề, Shophouse: Nếu diện tích đất từ 0m2 - 90m2 thì dtxd = dt đất, Nếu dt đất > 90m2: thì dtxd = 0.85 *dt đất
- Nếu không có các giá trị ở trên, mặc định dtxd = diện tích đất.

### 4. Đơn giá xây dựng (Cần bổ sung thêm sau khi scrape nhà liền kề)

- Nhà cấp 4: 4.000.000
- Nhà 1 tầng bê tông cốt thép: 6.000.000
- Nhà từ 2 - 6 tầng, bê tông cốt thép, không hầm: 9.500.000
- Nhà từ 2 - 6 tầng, bê tông cốt thép, có hầm: 11.000.000
- Nhà từ 7 tầng trở lên: 15.000.000
- Nhà biệt thự, không hầm: 15.000.000
- Nhà biệt thự, có hầm: 18.000.000

### 5. Chất lượng còn lại

Nếu tiêu đề bài đăng có các cụm từ "biệt thự, nhà liền kề, shophouse", lấy chất lượng còn lại tại trường "Tình trạng hoàn thiện": 
- Xây thô: 0.5
- Xây thô hoàn thiện mặt ngoài: 0.7
- Hoàn thiện cơ bản: 0.85
- Hoàn thiện đầy đủ: 1

Nếu là đất nền thì để trống, còn nếu là nhà thì mặc định là 0.85.